pkg/flatrpc/conn.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package flatrpc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	flatbuffers "github.com/google/flatbuffers/go"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/stats"
+)
+
+var (
+	statSent = stats.Create("rpc sent", "Outbound RPC traffic",
+		stats.Graph("traffic"), stats.Rate{}, stats.FormatMB)
+	statRecv = stats.Create("rpc recv", "Inbound RPC traffic",
+		stats.Graph("traffic"), stats.Rate{}, stats.FormatMB)
+)
+
+type Serv struct {
+	Addr *net.TCPAddr
+	ln   net.Listener
+}
+
+func ListenAndServe(addr string, handler func(*Conn)) (*Serv, error) {
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		for {
+			conn, err := ln.Accept()
+			if err != nil {
+				if strings.Contains(err.Error(), "use of closed network connection") {
+					break
+				}
+				var netErr *net.OpError
+				if errors.As(err, &netErr) && !netErr.Temporary() {
+					log.Fatalf("flatrpc: failed to accept: %v", err)
+				}
+				log.Logf(0, "flatrpc: failed to accept: %v", err)
+				continue
+			}
+			go func() {
+				c := newConn(conn)
+				defer c.Close()
+				handler(c)
+			}()
+		}
+	}()
+	return &Serv{
+		Addr: ln.Addr().(*net.TCPAddr),
+		ln:   ln,
+	}, nil
+}
+
+func (s *Serv) Close() error {
+	return s.ln.Close()
+}
+
+type Conn struct {
+	conn net.Conn
+
+	sendMu  sync.Mutex
+	builder *flatbuffers.Builder
+
+	data    []byte
+	hasData int
```

Comment:
I wonder if it could have been easier to just use `io.Copy` instead of `io.ReadAtLeast`: block on reading the first 4 bytes (size), then just read the whole message size. Then we wouldn't need `lastMsg` and `hasData` (?)

What are the benefits of the current approach?

Reply:
With you approach we may need 4+ syscalls to read what we can read now with 1 syscall (size + data + subsequent message size+data, which may be already in the socket buffer).

Reply:
If there are N messages in the buffer, then now we'd `copy(c.data, c.data[c.lastMsg:c.hasData])` N-th of them N times before processing.

If we don't care about copying that much, was https://pkg.go.dev/bufio#Reader an option?

Reply:
The most important case is reading just 1 message (I would assume it will happen most of the time), Current code reads it with 1 syscall instead of 2.
I don't see how we can use bufio.Reader, flatbuffers needs full []byte slice as an input, it does not operate on io.Reader.